micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,269 @@
-        let ret = crate::esp32::esp_idf_svc::sys::EspError::convert(ret);
-        ret.map_err(|e| CameraError::CameraInitError(e.into()))
-    }
-    pub fn get_cam_frame(&self) -> Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t> {
-        let ptr = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_fb_get() })
-            as *mut crate::esp32::esp_idf_svc::sys::camera_fb_t;
-        if ptr.is_null() {
-            None
-        } else {
-            Some(ptr)
+}
+
+impl Camera for Esp32Camera {
+    fn get_image(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
+        let frame = Esp32CameraFrameBuffer::get().ok_or(CameraError::FailedToGetImage)?;
+        if frame.len() > buffer.capacity() {
+            return Err(CameraError::ImageTooBig(frame.len(), buffer.capacity()));
         }
+        let msg = camera::v1::GetImageResponse {
+            mime_type: "image/jpeg".to_string(),
+            image: frame.as_bytes(),
+        };
+        // safety: message must be encoded before the frame is dropped from scope
+        msg.encode(&mut buffer)
+            .map_err(CameraError::MessageEncodeError)?;
+
+        return Ok(buffer);
     }
-    pub fn return_cam_frame(
-        &self,
-        frame: Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t>,
-    ) {
-        if let Some(ptr) = frame {
-            unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_fb_return(ptr) }
+
+    fn render_frame(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
+        let frame = Esp32CameraFrameBuffer::get().ok_or(CameraError::FailedToGetImage)?;
+        if frame.len() > buffer.capacity() {
+            return Err(CameraError::ImageTooBig(frame.len(), buffer.capacity()));
         }
+        let msg = HttpBody {
+            content_type: "image/jpeg".to_string(),
+            data: frame.as_bytes().to_vec(),
+            ..Default::default()
+        };
+        // safety: message must be encoded before the frame is dropped from scope
+        msg.encode(&mut buffer)
+            .map_err(CameraError::MessageEncodeError)?;
+        return Ok(buffer);
     }
-    pub fn debug_print_fb(&self, frame: &Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t>) {
-        if let Some(ptr) = frame {
-            let ptr = ptr as &*mut crate::esp32::esp_idf_svc::sys::camera_fb_t;
-            unsafe {
-                println!();
-                info!("camera buf size {}", (*(*ptr)).len);
-                println!();
-                for i in 0..(*(*ptr)).len {
-                    print!("{:02X}", (*(*(*ptr)).buf.offset(i as isize)));
-                    if i > 0 && i % 80 == 0 {
-                        println!();
-                    }
-                }
-            }
-        }
+}
+
+impl Drop for Esp32Camera {
+    fn drop(&mut self) {
+        unsafe { esp_camera_deinit() };
+        let mut is_registered = CAMERA_ALREADY_REGISTERED.lock().unwrap();
+        *is_registered = true;
```

Comment:
I think this should be `false`

Reply:
yup, just found it. doing testing before merging stuff. 
Also realized that the failed initialization of camera actually does a drop, which calls `esp_camera_deinit`, so fixing that now. will confirm proper behavior before merging.

I'm doing things like adding, removing, testing multiple cameras on app.

Reply:
OK. If you want me to review that new logic just re-request review and I'll take a look.